pilot/pkg/networking/plugin/authz/rbac.go

```diff
@@ -0,0 +1,319 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package authz
+
+import (
+	"fmt"
+	"strings"
+
+	xdsapi "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	http_conn "github.com/envoyproxy/go-control-plane/envoy/config/filter/network/http_connection_manager/v2"
+	policy "github.com/envoyproxy/go-control-plane/envoy/config/rbac/v2alpha"
+	"github.com/envoyproxy/go-control-plane/envoy/type"
+
+	rbacproto "istio.io/api/rbac/v1alpha1"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/networking/plugin"
+	"istio.io/istio/pilot/pkg/networking/util"
+	"istio.io/istio/pilot/pkg/serviceregistry/kube"
+	"istio.io/istio/pkg/log"
+)
+
+const (
+	// RbacFilterName is the name for the Rbac filter.
+	// TODO(yangminzhu): Update once the final name is decided.
+	RbacFilterName = "rbac-authz"
+)
+
+// Plugin implements Istio RBAC authz
+type Plugin struct{}
+
+// NewPlugin returns an instance of the authz plugin
+func NewPlugin() plugin.Plugin {
+	return Plugin{}
+}
+
+// OnOutboundListener is called whenever a new outbound listener is added to the LDS output for a given service
+// Can be used to add additional filters on the outbound path
+func (Plugin) OnOutboundListener(in *plugin.InputParams, mutable *plugin.MutableObjects) error {
+	return nil
+}
+
+// OnInboundListener is called whenever a new listener is added to the LDS output for a given service
+// Can be used to add additional filters (e.g., mixer filter) or add more stuff to the HTTP connection manager
+// on the inbound path
+func (Plugin) OnInboundListener(in *plugin.InputParams, mutable *plugin.MutableObjects) error {
+	rbacConfigs, err := in.Env.IstioConfigStore.List(model.RbacConfig.Type, kube.IstioNamespace)
+	if err != nil {
+		return fmt.Errorf("failed to get rbacConfig %v", err)
+	}
+	if len(rbacConfigs) != 1 {
+		return fmt.Errorf("found %d rbacConfigs, only supports 1 rbacConfig", len(rbacConfigs))
+	}
+	configProto := rbacConfigs[0].Spec.(*rbacproto.RbacConfig)
+	// TODO(yangminzhu): Supports ON_WITH_INCLUSION and ON_WITH_EXCLUSION.
+	if configProto.Mode != rbacproto.RbacConfig_ON {
+		log.Infof("rbac plugin disabled by rbacConfig %v", *configProto)
+		return nil
+	}
+
+	if in.Node.Type != model.Sidecar {
+		// Only supports sidecar proxy for now.
+		return nil
+	}
+
+	for i := range mutable.Listener.FilterChains {
```

Comment:
What is the relationship between mutable.Listener.FilterChains and mutable.FilterChains? I saw only mutable.FilterChains is filled out here.

Reply:
Refactored to only use mutable.FilterChains.